packages/vite/src/node/server/hmr.ts

```diff
@@ -144,18 +144,25 @@ export function updateModules(
   let needFullReload = false
 
   for (const mod of modules) {
-    moduleGraph.invalidateModule(mod, invalidatedModules, timestamp, true)
-    if (needFullReload) {
-      continue
-    }
-
     const boundaries: { boundary: ModuleNode; acceptedVia: ModuleNode }[] = []
     const hasDeadEnd = propagateUpdate(mod, traversedModules, boundaries)
     if (hasDeadEnd) {
```

Comment:
> If a module is a dead end, it's not invalidated

@ArnaudBarre Does it mean even though Vite perform a full reload but still could not get the edited content?

Reply:
I'm not sure because I'm surprised not test failed. But because the dev server has some in-memory cache, we need to be sure to invalidate this on file change. But maybe this is done elsewhere. I need to dig in again on this code to understand why we invalidate that much in the first place actually

Reply:
Probably because of this code

https://github.com/vitejs/vite/blob/447df7cba987b30e3621076a74e2227f8232f64a/packages/vite/src/node/server/index.ts#L540-L546

https://github.com/vitejs/vite/blob/447df7cba987b30e3621076a74e2227f8232f64a/packages/vite/src/node/server/moduleGraph.ts#L109-L118

`moduleGraph.onFileChange(file)` would delete the memory cache in the following code

https://github.com/vitejs/vite/blob/447df7cba987b30e3621076a74e2227f8232f64a/packages/vite/src/node/server/moduleGraph.ts#L131-L148